File: Tests/MozzCoreTests/LatestReleaseTests.swift
import Testing
@testable import MozzCore

@Suite("Latest release ordering")
struct LatestReleaseTests {
    private func r(_ year: Int?, _ addedAt: Double? = nil, _ title: String = "x") -> ReleaseRecency {
        ReleaseRecency(year: year, addedAt: addedAt, title: title)
    }

    @Test("A later year wins")
    func laterYearWins() {
        #expect(LatestRelease.isNewer(r(2024), than: r(2019)))
        #expect(!LatestRelease.isNewer(r(2019), than: r(2024)))
    }

    @Test("A dated release outranks an undated one, in both directions")
    func datedBeatsUndated() {
        #expect(LatestRelease.isNewer(r(1998), than: r(nil)))
        #expect(!LatestRelease.isNewer(r(nil), than: r(1998)))
    }

    @Test("Same year falls through to when the library saw it")
    func sameYearUsesAddedAt() {
        #expect(LatestRelease.isNewer(r(2024, 200), than: r(2024, 100)))
        #expect(!LatestRelease.isNewer(r(2024, 100), than: r(2024, 200)))
    }

    @Test("With neither year nor date the title decides, so the order is stable")
    func fallsBackToTitle() {
        #expect(LatestRelease.isNewer(r(nil, nil, "Aaa"), than: r(nil, nil, "Bbb")))
        #expect(!LatestRelease.isNewer(r(nil, nil, "Bbb"), than: r(nil, nil, "Aaa")))
    }

    @Test("Picks the newest from a list")
    func picksNewest() {
        let releases = [r(2019, nil, "old"), r(2026, nil, "new"), r(2022, nil, "middle")]
        #expect(LatestRelease.newestIndex(releases) == 1)
    }

    @Test("An empty list has no newest release")
    func emptyHasNone() {
        #expect(LatestRelease.newestIndex([]) == nil)
    }

    @Test("A discography with no years at all still has a stable answer")
    func undatedDiscographyIsStable() {
        // The case that actually happens: a self-hosted server returns a whole
        // artist with no year on anything. The answer must not depend on the
        // order the rows arrived in, or two clients disagree.
        let a = [r(nil, nil, "Bbb"), r(nil, nil, "Aaa"), r(nil, nil, "Ccc")]
        let b = [r(nil, nil, "Ccc"), r(nil, nil, "Bbb"), r(nil, nil, "Aaa")]
        #expect(a[LatestRelease.newestIndex(a)!].title == "Aaa")
        #expect(b[LatestRelease.newestIndex(b)!].title == "Aaa")
    }
}

File: clients/desktop-tests/LatestReleaseTests.cs
using Mozz.Desktop.Core;
using Mozz.Desktop.ViewModels;
using Xunit;

namespace Mozz.Desktop.Tests;

/// <summary>
/// Pins the desktop's "Latest Release" ordering to the shared core's.
/// </summary>
/// <remarks>
/// The rule lives in <c>Sources/MozzCore/LatestRelease.swift</c> and is mirrored
/// in <see cref="MediaDetailFormatting"/> because it is a pure function over data
/// the client already holds, and a round trip per album would be absurd for a
/// shelf. Mirroring is only safe if it is checked, so these cases are the same
/// cases as <c>Tests/MozzCoreTests/LatestReleaseTests.swift</c>. If someone
/// changes the rule on one side, this fails rather than letting the phone and the
/// desktop quietly name different records as an artist's latest.
/// </remarks>
public class LatestReleaseTests
{
    private static Album A(int? year, double? addedAt = null, string title = "x") =>
        new(Id: 0, RemoteId: title, ServerId: "srv", Title: title, ArtistName: "a",
            ArtistRemoteId: "ar", Year: year, TrackCount: 10, ArtworkKey: null,
            AddedAt: addedAt);

    [Fact]
    public void ALaterYearWins()
    {
        Assert.Equal("new", MediaDetailFormatting.LatestRelease(
            new[] { A(2019, title: "old"), A(2024, title: "new") })!.Title);
    }

    [Fact]
    public void ADatedReleaseOutranksAnUndatedOne_InBothDirections()
    {
        Assert.True(MediaDetailFormatting.IsNewer(A(1998), A(null)));
        Assert.False(MediaDetailFormatting.IsNewer(A(null), A(1998)));
    }

    [Fact]
    public void SameYearFallsThroughToWhenTheLibrarySawIt()
    {
        Assert.True(MediaDetailFormatting.IsNewer(A(2024, 200), A(2024, 100)));
        Assert.False(MediaDetailFormatting.IsNewer(A(2024, 100), A(2024, 200)));
    }

    [Fact]
    public void WithNeitherYearNorDateTheTitleDecides()
    {
        Assert.True(MediaDetailFormatting.IsNewer(A(null, null, "Aaa"), A(null, null, "Bbb")));
        Assert.False(MediaDetailFormatting.IsNewer(A(null, null, "Bbb"), A(null, null, "Aaa")));
    }

    [Fact]
    public void AnEmptyListHasNoLatestRelease()
    {
        Assert.Null(MediaDetailFormatting.LatestRelease(System.Array.Empty<Album>()));
    }

    [Fact]
    public void AnUndatedDiscographyStillHasAStableAnswer()
    {
        // The case that actually happens: a self-hosted server returns a whole
        // artist with no year on anything. The answer must not depend on the
        // order the rows arrived in, or two clients disagree about the same
        // library.
        var forwards = new[] { A(null, null, "Bbb"), A(null, null, "Aaa"), A(null, null, "Ccc") };
        var backwards = new[] { A(null, null, "Ccc"), A(null, null, "Bbb"), A(null, null, "Aaa") };
        Assert.Equal("Aaa", MediaDetailFormatting.LatestRelease(forwards)!.Title);
        Assert.Equal("Aaa", MediaDetailFormatting.LatestRelease(backwards)!.Title);
    }
}
